python/pyxet/pyxet/util.py

```diff
@@ -150,19 +159,46 @@ def _path_join(fs, path, *paths):
     if fs.protocol == 'file':
         return os.path.join(path, *paths)
     else:
-        return '/'.join([path] + list(map(lambda p: p.strip('/'), paths)))
+        return '/'.join([path] + [p.strip('/') for p in paths if len(p) != 0])
 
 
 def _path_dirname(fs, path):
     if fs.protocol == 'file':
         return os.path.dirname(path)
     else:
         return '/'.join(path.split('/')[:-1])
+    
+def _path_normalize(fs, path):
+    if fs.protocol == 'file':
+        return os.path.abspath(path)
+    elif path != '/':
+        return path.rstrip('/')
+    else:
+        return path
 
-
-def _copy(source, destination, recursive=True, _src_fs=None, _dest_fs=None):
+def _parse_and_sanitize_src_and_dest_path(source, destination):
     src_fs, src_path = _get_fs_and_path(source)
     dest_fs, dest_path = _get_fs_and_path(destination)
+
+    dest_isdir = _isdir(dest_fs, dest_path)
+    # Our target is an existing directory and src is not a wildcard copy
+    # i.e. we are getting cp src/some/path to dest/some/where
+    # but dest/some/where exists as a directory
+    # So we will need to make the dest dest/some/where/path
+    if dest_isdir and '*' not in src_path:
+        # Split up the final component from source path and add it
+        # to the destination.
+        # We get final component from 'source' instead of the 
+        # normalized 'src_path' because if 'source' ends with 
+        # '/' or '\', the desired behavior is to copy what's 
+        # under that directory, and to skip 'source' itself.
+        _, final_source_component = _path_split(src_fs, source)
+        dest_path = _path_join(dest_fs, dest_path, final_source_component)
+    return src_fs, src_path, dest_fs, dest_path
+
+def _build_src_and_dest_list(source, destination, recursive=False, _src_fs=None, _dest_fs=None):
```

Comment:
This function seems to either perform the action or build the lists depending on various things.  I think we should either make it do one thing or the other, or change the name to reflect this.  
- if src and dest is a xet, then it does the copy, otherwise it returns the copy list. 
- it creates all the directories needed, but then returns the list of files to be copied.